Nix superfluous tag checkout.

kriskowal opened the pull request:
Tests pass when I remove what appears to be a superfluous checkout of the version timestamp tag for the current project’s idl/meta.json. I am not sure this is correct, since I’m not sure why this was here in the first place, but my intuition is that each individual service fetch is independent, and does not depend on this initial checkout.

Is there a test I can add that would illustrate that this is a breaking change?

The purpose of this review is to ask @raynos @malandrew what the intended purpose of this line was, and whether I can safely remove it.

The reason I’m asking is that this would mitigate a bug I introduced when I released a change that stopped pushing a timestamp tag for every published commit. I may have to revert that change and backfill tags, but if they are unnecessary, this alternative is better.

I may be able to rewrite the relevant behavior of `idl` using git plumbing commands, but until then I’m looking for small fixes. From first principles, I believe version tags are unnecessary.

kriskowal commented:
Reading further and commenting preconditions, all commands in IDL start with fetching and checking out the registry's "origin/master", which I think is what we always want before "fetch" and "update", which convinces me that this is okay to land. Let me know if you agree.

kriskowal commented:
Followed up with a commit that removes the `fetchFromMeta` function entirely, so `fetch` without arguments is equivalent to `update`. That tests pass and that this is my intuition for the correct behavior of fetch, convinces me that the prior behavior checking out the version tag for each service was superfluous.

Ready for review to land.

andrewdeandrade commented:
LGTM. nice comments, sorry Jake and I didn't put them in when we originally worked on it.

kriskowal commented:
Prashant and I puzzled out the behavior. I noticed that the first-fetch behavior differs from update in that it fetches from the snapshot at the time of the last update, to ensure that the idl directory consists of a consistent cross-sectional snapshot of the idl registry at that commit, on the theory that there may be inter-service dependencies. This has the surprising result that it’s possible for `fetch` to obtain a back-dated version, which would then get immediately replaced by the next `update`.

I will have to look more closely at this tomorrow, but I do think it’s a simpler and less surprising if `every` fetch advances the idl registry to a cross-sectional snapshot of the registry’s `origin/master` instead of the back-dated version. To that end, we will want all fetches to also be updates, which may require further work.

kriskowal commented:
I’ve removed mutual recursion between `fetch` and `update`, by adding `fetchOneService`, which is self-recursive but only for collected dependencies. I did not realize that we had dependency analysis idl does not depend on the thriftrw-node parser. It uses a RegExp, which gives me these horrible chills.

I also moved the code that breaks recursion cycles to the top of `fetchOneService`, because this isn’t my first rodeo.

In any case, this made it possible to line up `fetch` and `update` so that `fetch` always updates existing services first, ensuring that both fetch and update synchronize the entire `idl/` directory with "origin/master" of the registry.

This is a small change in behavior that makes `idl` less surprising and gets rid of the problem with version tags.

kriskowal commented:
To be clear, the RegExp is the least of our problems for dependency analysis. This feature has sharp edges and won’t work if one dependency is not at the root of its containing service directory. Where service roots begin and sub-service directories begin is fundamentally unknowable to the idl tool for services that have not already been fetched.

That problem remains even if we replace getDependencies with a proper implementation based on the Thrift IDL AST parser.

kriskowal commented:
In the interest of unbreaking idl in production, I’m landing this based on Andrew’s review, though I would appreciate further reviews.
